Added FilterTrace and filtered select1-5.test

dniu16 opened the pull request:
Added FilterTrace, a class that filter out desired traces from a trace file, and its target filter-trace.

Instruction included in junit/README.md. Takes in a list of skip keywords. Queries that contain any keyword will be tagged as "skip" while the others will have the hash updated with postgresql. When TracefileTest is called, queries with the "skip" tag will be ignored.

Added filtered select1-5.test (queries without CASE, BETWEEN, WHERE, abs(, and WHEN in them).

select1.test: 14/97 failed
Failed query: Line 860, 1019, 1783, 3563, 4094, 4608, 5199, 5608, 6081, 6603, 7658, 9635, 10963, 11317

select2.test: 82 tests successful

select3.test: 157 tests successful

select4.test: 1043 tests successful, most of the tests are insert

select5.test: 704 tests successful, no query all insert

apavlo commented on script/testing/junit/src/GenerateTrace.java:
```diff
@@ -24,7 +25,7 @@
     public static final String SEPARATION = "----";
     public static final String DEST_DIR = "traces";
-    public static final String DEST_NAME = "output.test";
+    public static final String DEST_NAME = "output.txt";
```
Actually, you should probably pass this in as an argument.

dniu16 commented on script/testing/junit/src/TracefileTest.java:
```diff
@@ -91,31 +90,34 @@ public void setUp() throws Throwable {
                 // case for query statements
-                if(mog.queryResults.get(0).contains("values")){
+                String parsed_hash;
+                if(mog.queryResults.size()==0 || (!mog.queryResults.get(0).contains("values"))) {
```
Yeah if(mog.queryResults.size()==0 || (!mog.queryResults.get(0).contains(constants.VALUES))) handles the no-hash case

apavlo commented on script/testing/junit/src/moglib/constants.java:
```diff
@@ -0,0 +1,14 @@
+package moglib;
+
+public class constants {
```
Java naming convention is that file + class names should be uppercase

apavlo commented:
@dniu16 Please also incorporate @jkosh44's updates to the README from #1072 into this PR.

jkosh44 commented on script/testing/junit/src/FilterTrace.java:
```diff
@@ -0,0 +1,154 @@
+            mog.comments.clear();
+            mog.queryResults.clear();
+        }
+        writer.close();
```
I'm not sure if you're familiar with try-with-resources (https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html), but it tends to be a safer/cleaner way to manage closable resources. It may not be needed for a short lived script like this, but it's handy to know.

jkosh44 commented on script/testing/junit/src/GenerateTrace.java:
```diff
@@ -61,25 +59,16 @@ public static void main(String[] args) throws Throwable {
-                md.update(resultString.getBytes());
-                byte[] byteArr = md.digest();
-                String hash = MogUtil.bytesToHex(byteArr).toLowerCase();
+                String hash = TestUtility.getHashFromDb(res);
```
Something I keep running into is that if the number of decimal points are different in postgres and terrier I'll get the wrong hash. May be worthwhile to include it in the README somewhere (not sure what the fix is though, maybe if possible to change the precision of the postgres responses. Or maybe in the future we can add some rounding before hashing).

dniu16 commented:
1. I talked with Poojita and added support for checking expected number of results. When using GenerateTrace, if an expected number of results is given in the comment for a query, then in TracefileTest in addition to checking the hash, we would also check the number of query results. Instruction is in junit/README.md. 

2. Added several test_broken files:
constraints.test_broken (constraint file per Poojita's request): She was going to talk with Wuwen about the error. 

update.test_broken (additional update tests added by me): There were rounding errors: 
 Line:88 | Hash:d6eb3cbabeb480aef68da6106902c619 ✘ Failure at Line 88: 
08-17-2020 06:09:13,574 [test_server.py:181] INFO : [java] │        │        SELECT * FROM update4
08-17-2020 06:09:13,574 [test_server.py:181] INFO : [java] │        │        Query expected 9 results, got 9 results
08-17-2020 06:09:13,574 [test_server.py:181] INFO : [java] │        │        [1, 1.000000, 1, 3, 3.000000, 3, 2, 4.000000, 2]
08-17-2020 06:09:13,574 [test_server.py:181] INFO : [java] │        │        expected: <d6eb3cbabeb480aef68da6106902c619> but was: <1c8293b7dd6abf3bf901da789549d7b9>

select.test_broken (previously select.test): In addition to removing the COUNT(DISTINCT) query there was another error: line 140
query I nosort
SELECT a FROM foo_one WHERE a = '11223344556677889900112233'
----
1 values hashing to a8acc77b83bcfd87fe1d2136bc45d826
query got 0 results

3. It turns out the reason I got invalid as column type was because the float-handling part was missing in Wan's library, so I added the float case in MogDb.

apavlo commented:
@dniu16 Why add the broken file to this PR? Shouldn't Poojita add it to her PR?

apavlo commented:
@dniu16 Why add the broken file to this PR? Shouldn't @Poojita-Raj add it to her PR?

dniu16 commented:
Sure I'll remove constraints.test_broken and send the file to her.

apavlo commented on test/integration/CMakeLists.txt:
```diff
@@ -1 +0,0 @@
-ADD_TERRIER_TESTS()
```
Restore these deleted files

apavlo commented on script/testing/junit/src/TestUtility.java:
```diff
@@ -127,4 +137,43 @@ public void setValues(PreparedStatement pstmt, int[] values) throws SQLException
+     * @return list of integers that contains start line numbers
+     * @throws IOException
+     */
+    public static List<Integer> getQueryLineNum(File input) throws IOException {
+        BufferedReader bf = new BufferedReader(new FileReader(input));
+        List<Integer> res = new ArrayList<>();
+        String line;
+        int counter = 0;
+        while (null != (line = bf.readLine())){
+            counter++;
+            if(line.startsWith("query") || line.startsWith("statement")){
+                res.add(counter);
+            }
+        }
+        return res;
+    }
```
Is this function only needed by `TraceFileTest`? If yes, then put it inside that class.

apavlo commented on script/testing/junit/src/TracefileTest.java:
```diff
@@ -56,31 +47,30 @@ public void setUp() throws Throwable {
+            if (mog.queryFirstLine.equals(Constants.STATEMENT_ERROR)||mog.queryFirstLine.equals(Constants.STATEMENT_OK)) {
                 Statement statement = null;
                 Executable exec = null;
-                String testName = "Line:" + num + " | Expected " + mog.status;
+                String testName = "Line:" + num + " | Expected " + mog.queryFirstLine;
                 try {
                     statement = conn.createStatement();
                     statement.execute(cur_sql);
-                    if(mog.status.equals(ERROR)){
+                    if(mog.queryFirstLine.equals(Constants.STATEMENT_ERROR)){
                         String message = "Failure at Line " + num + ": Expected failure but success"  + "\n " + cur_sql;
                         exec = () -> check2(message);
                     }else{
                         exec = () -> assertEquals(true, true);
                     }
                 }
                 catch (Throwable e) {
-                    if(mog.status.equals(OK)){
+                    if(mog.queryFirstLine.equals(Constants.STATEMENT_OK)){
                         String message = "Failure at Line " + num + ": Expected success but failure"  + "\n " + cur_sql;
                         exec = () -> check2(message);
                     }else{
```
The `generateTest` function is getting too large. Split out the two main if-clause branches into separate functions: `executeQueryNoResults` and `executeQueryWithResults`

apavlo commented on script/testing/junit/src/TracefileTest.java:
```diff
@@ -128,15 +142,20 @@ public void setUp() throws Throwable {
      * @throws Exception
      */
-    public static void check(String hash1, String hash2, int n, String sql) throws Exception {
+    public static void check(String hash1, String hash2, String message,
```
Can you rename this function to be more descriptive than just `check`?

apavlo commented on script/testing/junit/src/TracefileTest.java:
```diff
@@ -149,44 +168,4 @@ public static void check(String hash1, String hash2, int n, String sql) throws E
     public static void check2(String mes) throws Exception {
```
Likewise, rename this to be something other than `check2`. I think it should be `checkAlwaysFail`

apavlo commented on script/testing/junit/traces/insert.sql:
```diff
@@ -1,84 +0,0 @@
-# 1 tuple insert, with no column specification.
```
Why was this file deleted?

apavlo commented on script/testing/junit/traces/functions.sql:
```diff
@@ -1,15 +0,0 @@
-# Initialize the database and table for testing
```
Why was this file deleted?

apavlo commented on script/testing/junit/traces/select.sql:
```diff
@@ -1,37 +0,0 @@
-# SELECT With Duplicate Columns Produces Zero Results
```
Why was this file deleted?
